File: python3/binary_tree_zigzag_level_order_traversal.py
from model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""


class Solution:

    def getDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        l = self.getDepth(root.left) + 1
        r = self.getDepth(root.right) + 1

        return max(l, r)

    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []

        ans = []
        maxdep = self.getDepth(root)

        for _ in range(maxdep):
            ans.append([])

        stack = [(root, 0)]

        while stack:
            node, level = stack.pop(0)
            ans[level].append(node.val)
            if node.left:
                stack.append((node.left, level + 1))
            if node.right:
                stack.append((node.right, level + 1))

        direct = "l"
        for j in range(maxdep):
            if direct == "l":
                direct = "r"
                continue
            ans[j] = ans[j][::-1]
            direct = "l"

        return ans

File: golang/binary_tree_zigzag_level_order_traversal.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

func zigzagLevelOrder(root *model.TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	ans := [][]int{}
	lines := [][]int{}
	stack := []*model.TreeNode{root}
	for len(stack) > 0 {
		line := []int{}
		temp := []*model.TreeNode{}

		for i := 0; i < len(stack); i++ {
			line = append(line, stack[i].Val)
			if stack[i].Left != nil {
				temp = append(temp, stack[i].Left)
			}
			if stack[i].Right != nil {
				temp = append(temp, stack[i].Right)
			}
		}
		stack = temp
		lines = append(lines, line)
	}

	dir := "+"
	for _, arr := range lines {
		if dir == "+" {
			ans = append(ans, arr)
		} else {
			rev := []int{}
			for i := len(arr) - 1; i >= 0; i-- {
				rev = append(rev, arr[i])
			}
			ans = append(ans, rev)
		}

		if dir == "+" {
			dir = "-"
		} else {
			dir = "+"
		}
	}

	return ans
}
